worker/src/main/scala/com/lucidchart/piezo/ConnectionProvider.scala

```diff
@@ -1,25 +1,134 @@
 package com.lucidchart.piezo
 
-import org.quartz.utils.HikariCpPoolingConnectionProvider
+import java.net.UnknownHostException
+import java.sql.{Connection, SQLTransientConnectionException}
 import java.util.Properties
+import java.util.concurrent.TimeUnit
+import org.quartz.utils.HikariCpPoolingConnectionProvider
 import org.slf4j.LoggerFactory
+import scala.annotation.tailrec
+import scala.collection.mutable.LinkedHashMap
 
 class ConnectionProvider(props: Properties) {
+
+  private class Pool(val ip: String) {
+    val connectionProvider: Option[HikariCpPoolingConnectionProvider] = createNewConnectionProvider()
+    logger.info(s"Initialized Db connection pool for ${jdbcURL}")
+    // Hikari takes about a second to add connections to the connection pool
+    // We are now going to warm-up connectionPool(with timelimit of 2500ms)
+    connectionProvider.map(warmUpCP)
+  }
+
   val logger = LoggerFactory.getLogger(this.getClass)
   private val dataSource = props.getProperty("org.quartz.jobStore.dataSource")
-  private val provider = if(dataSource != null) {
-    Some(new HikariCpPoolingConnectionProvider(
-      props.getProperty("org.quartz.dataSource." + dataSource + ".driver"),
-      props.getProperty("org.quartz.dataSource." + dataSource + ".URL"),
-      props.getProperty("org.quartz.dataSource." + dataSource + ".user"),
-      props.getProperty("org.quartz.dataSource." + dataSource + ".password"),
-      props.getProperty("org.quartz.dataSource." + dataSource + ".maxConnections").toInt,
-      props.getProperty("org.quartz.dataSource." + dataSource + ".validationQuery")
-    ))
-  } else {
-    logger.info("No job store found in config")
-    None
-  }
-
-  def getConnection = provider.get.getConnection
+  private val jdbcURL = if (dataSource != null) props.getProperty("org.quartz.dataSource." + dataSource + ".URL") else null
+  private val detectIpAddressFailover = if (dataSource != null) props.getProperty("org.quartz.dataSource." + dataSource + ".ipFailover") == "true" else false
+  // Removes "jdbc:mysql://" prefix and ":{port}..." suffix
+  private val dataSourceHostname = if (jdbcURL != null) jdbcURL.replace("jdbc:mysql://", "").split(":")(0) else null
+
+  // Class for storing the ip address of the host, along with an expiration date
+  private case class CachedIpWithExpiration(ip: String, expiration: Long)
+  // Cache for ip address and its expiration for a host
+  private val cachedIpWithExpiration: LinkedHashMap[String, CachedIpWithExpiration] = new LinkedHashMap()
```

Comment:
I think we need some synchronization on this, since multiple threads  could be  reading and writing to it in parallel. 

We should either use https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Collections.html#synchronizedMap(java.util.Map) or add synchronization around places that use this.

Also, should we use a subclass that overrides `removeEldestEntry` so that we can ensure the size of the map doesn't get too big?

Reply:
I like the idea of using the static `Collections.synchronizedMap(...)` constructor so we don't have to keep track of adding `syncrhonized` blocks everywhere. As for the `removeEldestEntry`, we should override it whenever the size of the map exceeds 1 so that we don't have the map consuming so much memory over time